source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.cc

```diff
@@ -0,0 +1,181 @@
+#include "extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.h"
+
+#include <atomic>
+#include <chrono>
+
+#include "envoy/config/filter/http/adaptive_concurrency/v2alpha/adaptive_concurrency.pb.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/runtime/runtime.h"
+#include "envoy/stats/stats.h"
+
+#include "common/common/cleanup.h"
+#include "common/protobuf/protobuf.h"
+#include "common/protobuf/utility.h"
+
+#include "extensions/filters/http/adaptive_concurrency/concurrency_controller/concurrency_controller.h"
+
+#include "absl/synchronization/mutex.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace AdaptiveConcurrency {
+namespace ConcurrencyController {
+
+GradientControllerConfig::GradientControllerConfig(
+    const envoy::config::filter::http::adaptive_concurrency::v2alpha::GradientControllerConfig&
+        proto_config)
+    : min_rtt_calc_interval_(std::chrono::milliseconds(
+          DurationUtil::durationToMilliseconds(proto_config.min_rtt_calc_params().interval()))),
+      sample_rtt_calc_interval_(std::chrono::milliseconds(DurationUtil::durationToMilliseconds(
+          proto_config.concurrency_limit_params().concurrency_update_interval()))),
+      max_concurrency_limit_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(
+          proto_config.concurrency_limit_params(), max_concurrency_limit, 1000)),
+      min_rtt_aggregate_request_count_(
+          PROTOBUF_GET_WRAPPED_OR_DEFAULT(proto_config.min_rtt_calc_params(), request_count, 50)),
+      max_gradient_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(proto_config.concurrency_limit_params(),
+                                                    max_gradient, 2.0)),
+      sample_aggregate_percentile_(PROTOBUF_PERCENT_TO_ROUNDED_INTEGER_OR_DEFAULT(
+                                       proto_config, sample_aggregate_percentile, 1000, 500) /
+                                   1000.0) {}
+
+GradientController::GradientController(GradientControllerConfigSharedPtr config,
+                                       Event::Dispatcher& dispatcher, Runtime::Loader&,
+                                       const std::string& stats_prefix, Stats::Scope& scope)
+    : config_(std::move(config)), dispatcher_(dispatcher), scope_(scope),
+      stats_(generateStats(scope_, stats_prefix)), deferred_limit_value_(1), num_rq_outstanding_(0),
+      concurrency_limit_(1), latency_sample_hist_(hist_fast_alloc(), hist_free) {
+  min_rtt_calc_timer_ = dispatcher_.createTimer([this]() -> void {
+    absl::MutexLock ml(&sample_mutation_mtx_);
+    enterMinRTTSamplingWindow();
+  });
+
+  sample_reset_timer_ = dispatcher_.createTimer([this]() -> void {
+    if (inMinRTTSamplingWindow()) {
```

Comment:
Sorry can you explain to me again why we need this check? Why can't we just arm the timer when we finish with the minRTT calculation? Can we do that and just replace this with an assert that we don't get here during minRTT calculation?

Reply:
This protects us from waiting on the mutex in the case where the sample timer is reset after `resetSampleWindow()` and the minRTT calculation kicks in before the sample reset timer fires. Without the check, we'd wait on the mutex until the minRTT calculation is finished, which could take a non-trivial amount of time due to waiting on some number of requests to come in.

I just didn't want to block the dispatcher thread. If I'm misguided in any of that logic and should rip it out, happy to do so.

Reply:
Ah OK I see the case where this happens. Once option would be to just cancel/disable the reset timer in this case. If you decide to keep this can you add more comments on when/why this happens?